### PR TITLE
add Canvas.triangle and Canvas.trianglefill

### PR DESCRIPTION
**Pull request by trelemar**

### CODE REVIEW

**Comment by trelemar** [comment]
I know I still need to add documentation but after that I'll mark it ready.

**Comment by BenStigsen** [comment]
I'd prefer calling this triangle honestly.

**Comment by avivbeeri** [comment]
I agree, `Canvas.triangle` and `Canvas.trianglefill` would match the existing API for circle and ellipse (rectangle is the odd one out, let's not talk about that.......)

**Comment by trelemar** [comment]
Haha, I took reference from the oddball while working on this. I agree.

File: src/engine.c [hunk 1]
```diff
@@ -869,6 +869,79 @@ ENGINE_ellipse(ENGINE* engine, int64_t x0, int64_t y0, int64_t x1, int64_t y1, u
...
+
+  for (int scanlineY = y0; scanlineY <= y1; scanlineY++) {
+    ENGINE_line(engine, curx0, scanlineY, curx1, scanlineY, c, 1);
+    curx0 += invslope0;
+    curx1 += invslope1;
+  }
+}
+
+internal void
+ENGINE_trianglefillFlatTop(ENGINE* engine, float x0, float y0, float x1, float y1, float x2, float y2, uint32_t c) {
```
Comment (avivbeeri) [inline src/engine.c:894]: ENGINE_trianglefillFlatBottom and ENGINE_trianglefillFlatTop look so similar, is there no way to combine them into a single method? (I don't fully grasp the triangle algorithm at the moment)

File: src/engine.c [hunk 2]
```diff
@@ -869,6 +869,79 @@ ENGINE_ellipse(ENGINE* engine, int64_t x0, int64_t y0, int64_t x1, int64_t y1, u
...
+  } else if (y0 == y1) {
+    ENGINE_trianglefillFlatTop(engine, x0, y0, x1, y1, x2, y2, c);
+  } else {
+    float x3 = (x0 + ((y1 - y0) / (y2 - y0)) * (x2 - x0));
+    ENGINE_trianglefillFlatBottom(engine, x0, y0, x1, y1, x3, y1, c);
+    ENGINE_trianglefillFlatTop(engine, x1, y1, x3, y1, x2, y2, c);
+  }
+  ENGINE_print(engine, "p0", x0, y0, 0xff00ccff);
+  ENGINE_print(engine, "p1", x1, y1, 0xff00ccff);
+  ENGINE_print(engine, "p2", x2, y2, 0xff00ccff);
```
Comment (avivbeeri) [inline src/engine.c:942]: Don't forget to remove these, as they are likely for debugging.

**Review by avivbeeri — CHANGES_REQUESTED** [review]

File: src/engine.c (hunk 2 above)
Comment (trelemar) [reply src/engine.c:942]: Will do before marking this ready :+1:

File: src/modules/graphics.c [hunk 3]
```diff
@@ -128,6 +128,50 @@ CANVAS_ellipsefill(WrenVM* vm)
...
+  int64_t x2 = round(wrenGetSlotDouble(vm, 5));
+  int64_t y2 = round(wrenGetSlotDouble(vm, 6));
+
+  uint32_t c = round(wrenGetSlotDouble(vm, 7));
+  ENGINE_triangle(engine, x0, y0, x1, y1, x2, y2, c);
+}
+
+internal void
+CANVAS_trianglefill(WrenVM* vm)
+{
```
Comment (avivbeeri) [inline src/modules/graphics.c:155]: Subtle formatting thing, the braces in this file are in the wrong place.

We keep them on the same line as the preceding syntax in the DOME project

**Review by avivbeeri — CHANGES_REQUESTED** [review]

**Comment by avivbeeri** [comment]
This PR generally looks good, just a few nitpicks here and there on style.

Thank you so much for providing this implementation.

Don't forget to add yourself to the AUTHORS file :)

File: src/modules/graphics.c (hunk 3 above)
Comment (trelemar) [reply src/modules/graphics.c:155]: Fixed in a201791f71d355a08a97dc0183cb7359d044d7fa

File: docs/modules/graphics.md [hunk 4]
```diff
@@ -92,6 +92,12 @@ Resize the canvas to the given `width` and `height`, and reset the color of the
 If `c` isn't provided, we default to black.
 Resizing the canvas resets the "clipping region" to encompass the whole canvas, as if `Canvas.clip()` was called.
 
+#### `static triangle(x0: Number, y0: Number, x1: Number, y1: Number, x2: Number, y3: Number, c: Color) `
+Draw a triangle with vertices at (_x0, y0_), (_x1, y1_), (_x2, y2_) in the color _c_.
+
+#### `static trianglefill(x0: Number, y0: Number, x1: Number, y1: Number, x2: Number, y3: Number, c: Color) `
```
Comment (avivbeeri) [inline docs/modules/graphics.md:98]: We probably shouldn't have an `x2` and a `y3` ;) 

Let's make that `x2, y2`

Comment (trelemar) [reply docs/modules/graphics.md]: fixed in bf92c10

**Review by avivbeeri — APPROVED** [review]

**Pull request merged**